### ISSUE TITLE
AdamW is no longer correctly implemented as per the cited paper

### ISSUE THREAD

**Issue by Zyrin** [keras-team/keras#19051]
As far as I can tell the current implementation of AdamW has a regression and is now incorrect according to the paper (https://arxiv.org/abs/1711.05101) that is cited in the documentation. This happened after the introduction of weight decay for all optimizers by changing the implementation of AdamW to inherit from Adam. Thus it is now "Adam with L2 regularization" instead of "Adam with decoupled weight decay".

**Comment by MoritzMo123** [comment]
From my understanding you are right. The default implementations overwriting


The [BaseOptimizer](https://github.com/keras-team/keras/blob/v3.0.2/keras/optimizers/base_optimizer.py#L655) implements apply_weight_decay in an AdamW style with param - wd * lr *param.

```
def _apply_weight_decay(self, variables):
        if self.weight_decay is None:
            return
        for variable in variables:
            if self._use_weight_decay(variable):
                lr = ops.cast(self.learning_rate, variable.dtype)
                wd = ops.cast(self.weight_decay, variable.dtype)
                variable.assign(variable - variable * wd * lr)
```

However, this is overwritten by both the torch and TensorFlow base optimizers, that inherit from this base class.
The  [TensorFlow BaseOptimizer](https://github.com/keras-team/keras/blob/v3.0.2/keras/backend/tensorflow/optimizer.py#L78) again implements AdamW style param = param - wd * lr * param, which means the basic keras behavior is replicated on the TensorFlow side. Note that this is in line with TensorFlow 2.15 where this flavor of weight decay is also applied for all  [optimizers](https://github.com/keras-team/keras/blob/v2.15.0/keras/optimizers/optimizer.py#L881))
```
def _apply_weight_decay(self, variables):
        if self.weight_decay is None:
            return

        def distributed_apply_weight_decay(distribution, variables, **kwargs):
            def weight_decay_fn(variable):
                if self._use_weight_decay(variable):
                    lr = tf.cast(self.learning_rate, variable.dtype)
                    wd = tf.cast(self.weight_decay, variable.dtype)
                    variable.assign_sub(variable * wd * lr)

            for variable in variables:
                if isinstance(variable, backend.Variable):
                    variable = variable.value  # Convert to tf.Variable
                distribution.extended.update(
                    variable, weight_decay_fn, group=False
                )

        tf.__internal__.distribute.interim.maybe_merge_call(
            distributed_apply_weight_decay,
            self._distribution_strategy,
            variables,
        )
```

On the other hand, the [torch BaseOptimizer](https://github.com/keras-team/keras/blob/v3.0.2/keras/backend/torch/optimizers/torch_optimizer.py#L8) uses param = param - wd * param for all optimizers. **This includes ADAMW, which should have param = param - wd * lr * param.**

```
def _apply_weight_decay(self, variables):
      if self.weight_decay is None:
          return

      torch._foreach_mul_(
          [v.value for v in variables if self._use_weight_decay(v)],
          1 - self.weight_decay * self._get_current_learning_rate(),
      )
```

[Jax BaseOptimizer](https://github.com/keras-team/keras/blob/v3.0.2/keras/backend/jax/optimizer.py) is the only one not to overwrite this behavior, therefore sticking with the base optimizer implementation.

**Comment by Zyrin** [comment]
I'm coming from the tensorflow side, and the problem that I thought was there actually isn't. I misunderstood something. But at least a problem on the torch side was found that way. Shall I close this issue? Or do you want to use it for the torch issue?

**Comment by MoritzMo123** [comment]
> I'm coming from the tensorflow side, and the problem that I thought was there actually isn't. I misunderstood something. But at least a problem on the torch side was found that way. Shall I close this issue? Or do you want to use it for the torch issue?

Nope, you can close it. I made a mistake: `1 - self.weight_decay * self._get_current_learning_rate(),` is actually lr * wd, somehow it went by my when I wrote the answer... it is actually the same behavior for all torch, tf and Jax.
